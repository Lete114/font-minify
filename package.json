{
  "name": "web-font-minify",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "vite build",
    "preview": "vite build && vite preview"
  },
  "devDependencies": {
    "@sveltejs/vite-plugin-svelte": "^3.0.1",
    "postcss-nesting": "^12.0.2",
    "svelte": "^4.2.8",
    "unocss": "^0.58.3",
    "vite": "^5.0.8"
  },
  "dependencies": {
    "font-minify": "0.0.0-alpha.4",
    "mini-i18n": "1.0.0-beta.2",
    "msg-alert": "1.0.0-beta.4"
  }
}
