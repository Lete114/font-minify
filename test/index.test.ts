import fs from 'node:fs'
import { fileURLToPath } from 'node:url'
import { dirname, join } from 'node:path'
import crypto from 'node:crypto'
import type { Buffer } from 'node:buffer'
import { describe, expect, it } from 'vitest'
import minify from '../src/index'
import { version } from '../package.json'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

const TEST_STRING = `Font-Minify-${version} 你说，儿豁中国人不骗中国人`

function getHash(buffer: Buffer) {
  return crypto.createHash('sha256')
    .update(buffer)
    .digest('hex')
}

const numberFormatter = new Intl.NumberFormat('en', {
  maximumFractionDigits: 2,
  minimumFractionDigits: 2,
})

/**
 * calculation of size
 * @param { number } byte
 * @returns display size
 */
export function displaySize(byte: number) {
  const ONE_MILLION = 1000000
  const ONE_THOUSAND = 1000

  if (byte >= ONE_MILLION) {
    const size = `${numberFormatter.format(byte / ONE_MILLION)} MB`
    return size
  }

  return `${numberFormatter.format(byte / ONE_THOUSAND)} kB`
}

describe('ttf', () => {
  const fontPath = join(__dirname, '../fonts/SmileySans-Oblique.ttf')
  const buffer = fs.readFileSync(fontPath)

  it('ttf', async () => {
    const ext = 'ttf'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: ext,
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "e21b047b299ae8bba8d97bc5d371a6d84ca1ae22117f0fd4bcbd93772d33b547",
        "newSize": "6.54 kB",
        "originSize": "2.10 MB",
      }
    `)
  })

  it('ttf 2 woff', async () => {
    const ext = 'ttf'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'woff',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "08fcdc492bf789bdcbffc3f6c3183b41cd8cb287ba3257ad63e6476627c0ab6d",
        "newSize": "6.61 kB",
        "originSize": "2.10 MB",
      }
    `)
  })

  it('ttf 2 woff2', async () => {
    const ext = 'ttf'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'woff2',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "474ccea891777f6fbc64eaf043eb664a71eacf2da15874d93d2c0bd8c38182b0",
        "newSize": "3.43 kB",
        "originSize": "2.10 MB",
      }
    `)
  })

  it('ttf 2 eot', async () => {
    const ext = 'ttf'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'eot',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "49f8440f4e6cc17247874f8830c5cae2faf714e56b7da82ed1788b3525ca117b",
        "newSize": "6.74 kB",
        "originSize": "2.10 MB",
      }
    `)
  })
})

describe('otf', () => {
  const fontPath = join(__dirname, '../fonts/SmileySans-Oblique.otf')
  const buffer = fs.readFileSync(fontPath)

  it('otf 2 ttf', async () => {
    const ext = 'otf'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'ttf',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "516149469cdf1d81807d99246bb9b1ac3086fe2cae43065dbf98dbf42c22bb87",
        "newSize": "6.84 kB",
        "originSize": "1.63 MB",
      }
    `)
  })

  it('otf 2 woff', async () => {
    const ext = 'otf'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'woff',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "492d597f473caa99efcfe2ce569e5925d71ae19a2da40c0fab155fac0636be25",
        "newSize": "6.91 kB",
        "originSize": "1.63 MB",
      }
    `)
  })

  it('otf 2 woff2', async () => {
    const ext = 'otf'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'woff2',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "dc62333ab764d31696da2b8135abfda6fda01ea79bc27fbb5380b41d001092cb",
        "newSize": "3.56 kB",
        "originSize": "1.63 MB",
      }
    `)
  })

  it('otf 2 eot', async () => {
    const ext = 'otf'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'eot',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "df319b8eb2612d2cf4aa18a8fef75adf0a213427911c6480540a5467fba62285",
        "newSize": "7.04 kB",
        "originSize": "1.63 MB",
      }
    `)
  })
})

describe('eot', () => {
  const fontPath = join(__dirname, '../fonts/SmileySans-Oblique.eot')

  const buffer = fs.readFileSync(fontPath)
  it('eot', async () => {
    const ext = 'eot'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'eot',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "49f8440f4e6cc17247874f8830c5cae2faf714e56b7da82ed1788b3525ca117b",
        "newSize": "6.74 kB",
        "originSize": "2.07 MB",
      }
    `)
  })

  it('eot 2 ttf', async () => {
    const ext = 'eot'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'ttf',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "e21b047b299ae8bba8d97bc5d371a6d84ca1ae22117f0fd4bcbd93772d33b547",
        "newSize": "6.54 kB",
        "originSize": "2.07 MB",
      }
    `)
  })

  it('eot 2 woff', async () => {
    const ext = 'eot'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'woff',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "08fcdc492bf789bdcbffc3f6c3183b41cd8cb287ba3257ad63e6476627c0ab6d",
        "newSize": "6.61 kB",
        "originSize": "2.07 MB",
      }
    `)
  })

  it('eot 2 woff2', async () => {
    const ext = 'eot'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'woff2',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "474ccea891777f6fbc64eaf043eb664a71eacf2da15874d93d2c0bd8c38182b0",
        "newSize": "3.43 kB",
        "originSize": "2.07 MB",
      }
    `)
  })
})

describe('woff', () => {
  const fontPath = join(__dirname, '../fonts/SmileySans-Oblique.woff')
  const buffer = fs.readFileSync(fontPath)

  it('woff', async () => {
    const ext = 'woff'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: ext,
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "08fcdc492bf789bdcbffc3f6c3183b41cd8cb287ba3257ad63e6476627c0ab6d",
        "newSize": "6.61 kB",
        "originSize": "2.07 MB",
      }
    `)
  })

  it('woff 2 ttf', async () => {
    const ext = 'woff'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'ttf',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "e21b047b299ae8bba8d97bc5d371a6d84ca1ae22117f0fd4bcbd93772d33b547",
        "newSize": "6.54 kB",
        "originSize": "2.07 MB",
      }
    `)
  })

  it('woff 2 woff', async () => {
    const ext = 'woff'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'woff',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "08fcdc492bf789bdcbffc3f6c3183b41cd8cb287ba3257ad63e6476627c0ab6d",
        "newSize": "6.61 kB",
        "originSize": "2.07 MB",
      }
    `)
  })

  it('woff 2 eot', async () => {
    const ext = 'woff'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'eot',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "49f8440f4e6cc17247874f8830c5cae2faf714e56b7da82ed1788b3525ca117b",
        "newSize": "6.74 kB",
        "originSize": "2.07 MB",
      }
    `)
  })
})

describe('woff2', () => {
  const fontPath = join(__dirname, '../fonts/SmileySans-Oblique.ttf.woff2')
  const buffer = fs.readFileSync(fontPath)

  it('woff2', async () => {
    const ext = 'woff2'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: ext,
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "474ccea891777f6fbc64eaf043eb664a71eacf2da15874d93d2c0bd8c38182b0",
        "newSize": "3.43 kB",
        "originSize": "942.54 kB",
      }
    `)
  })

  it('woff2 2 ttf', async () => {
    const ext = 'woff2'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'ttf',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "5a372044143aca8b9188a49f9424e540d875c0af26b9bd56fb2c63c3cee7fcdd",
        "newSize": "6.54 kB",
        "originSize": "942.54 kB",
      }
    `)
  })

  it('woff2 2 woff', async () => {
    const ext = 'woff2'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'woff',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "8d661f2e87aba33856e0b65ddf69fa13f5f990e3f038568b5e8895589ed9061b",
        "newSize": "6.61 kB",
        "originSize": "942.54 kB",
      }
    `)
  })

  it('woff2 2 eot', async () => {
    const ext = 'woff2'
    const newBuffer = await minify({
      buffer,
      text: TEST_STRING,
      readOptions: {
        type: ext,
      },
      writeOptions: {
        type: 'eot',
      },
    })

    const data = {
      hash: getHash(newBuffer),
      originSize: displaySize(buffer.length),
      newSize: displaySize(newBuffer.length),
    }

    expect(data).toMatchInlineSnapshot(`
      {
        "hash": "1a30cfef36c6035cf21beb9d0c0913b7a4c52b38d9bbb2e51b6e71aa631e6433",
        "newSize": "6.74 kB",
        "originSize": "942.54 kB",
      }
    `)
  })
})
