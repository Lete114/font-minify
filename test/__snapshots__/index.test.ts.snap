// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`eot > eot 1`] = `
{
  "hash": "2750719ba61570365a121a2e5af958164db3cc07da51a92d89621974cd5126d6",
  "newSize": "6.83 kB",
  "originSize": "2.07 MB",
}
`;

exports[`eot > eot 2 ttf 1`] = `
{
  "hash": "3411e53f99d10f61099014e3411c3b07250c14911ce3228a972c7b86f528320e",
  "newSize": "6.63 kB",
  "originSize": "2.07 MB",
}
`;

exports[`eot > eot 2 woff 1`] = `
{
  "hash": "e575f399559aa7515331bcac1656b7129af19c0ffd0f6f3d40edf9ab542dec88",
  "newSize": "6.70 kB",
  "originSize": "2.07 MB",
}
`;

exports[`eot > eot 2 woff2 1`] = `
{
  "hash": "d674312638c499b7137a116536c1019cb8a263e16fe1f2561cf68316535cfde3",
  "newSize": "3.50 kB",
  "originSize": "2.07 MB",
}
`;

exports[`icon font > css 1`] = `
{
  "hash": "d148b1961a21a41c157dced63d86b29a2eee08b9044a0feb02e7e7df6fcb3d5a",
  "newSize": "4.93 kB",
  "originSize": "2.10 MB",
}
`;

exports[`icon font > html + css 1`] = `
{
  "hash": "d148b1961a21a41c157dced63d86b29a2eee08b9044a0feb02e7e7df6fcb3d5a",
  "newSize": "4.93 kB",
  "originSize": "2.10 MB",
}
`;

exports[`icon font > html 1`] = `
{
  "hash": "861a016c783b596ce167143deb3a922a7ef3d0a34d85ae90ebe959c12ced78af",
  "newSize": "4.45 kB",
  "originSize": "2.10 MB",
}
`;

exports[`otf > otf 2 eot 1`] = `
{
  "hash": "6acdca7e4cac5c098dd352def7f7dac6151dd5b56563e0510c0359df2c7a1fc1",
  "newSize": "7.16 kB",
  "originSize": "1.63 MB",
}
`;

exports[`otf > otf 2 ttf 1`] = `
{
  "hash": "4f34239d3e8848b4c15dbb8de3ff0a5a1cf59e5984bf6e9a31d50197264fdf3d",
  "newSize": "6.96 kB",
  "originSize": "1.63 MB",
}
`;

exports[`otf > otf 2 woff 1`] = `
{
  "hash": "cb390114bae304b8358c46e650c2cc01887c5c7cc26cec1995114ed2d1a548e2",
  "newSize": "7.03 kB",
  "originSize": "1.63 MB",
}
`;

exports[`otf > otf 2 woff2 1`] = `
{
  "hash": "0c571af6c677db43bfddea91616e26126d6ca22af92a2b9ef51a2c1868cb8c91",
  "newSize": "3.63 kB",
  "originSize": "1.63 MB",
}
`;

exports[`ttf > ttf 1`] = `
{
  "hash": "3411e53f99d10f61099014e3411c3b07250c14911ce3228a972c7b86f528320e",
  "newSize": "6.63 kB",
  "originSize": "2.10 MB",
}
`;

exports[`ttf > ttf 2 eot 1`] = `
{
  "hash": "2750719ba61570365a121a2e5af958164db3cc07da51a92d89621974cd5126d6",
  "newSize": "6.83 kB",
  "originSize": "2.10 MB",
}
`;

exports[`ttf > ttf 2 woff 1`] = `
{
  "hash": "e575f399559aa7515331bcac1656b7129af19c0ffd0f6f3d40edf9ab542dec88",
  "newSize": "6.70 kB",
  "originSize": "2.10 MB",
}
`;

exports[`ttf > ttf 2 woff2 1`] = `
{
  "hash": "d674312638c499b7137a116536c1019cb8a263e16fe1f2561cf68316535cfde3",
  "newSize": "3.50 kB",
  "originSize": "2.10 MB",
}
`;

exports[`woff > woff 1`] = `
{
  "hash": "e575f399559aa7515331bcac1656b7129af19c0ffd0f6f3d40edf9ab542dec88",
  "newSize": "6.70 kB",
  "originSize": "2.07 MB",
}
`;

exports[`woff > woff 2 eot 1`] = `
{
  "hash": "2750719ba61570365a121a2e5af958164db3cc07da51a92d89621974cd5126d6",
  "newSize": "6.83 kB",
  "originSize": "2.07 MB",
}
`;

exports[`woff > woff 2 ttf 1`] = `
{
  "hash": "3411e53f99d10f61099014e3411c3b07250c14911ce3228a972c7b86f528320e",
  "newSize": "6.63 kB",
  "originSize": "2.07 MB",
}
`;

exports[`woff > woff 2 woff 1`] = `
{
  "hash": "e575f399559aa7515331bcac1656b7129af19c0ffd0f6f3d40edf9ab542dec88",
  "newSize": "6.70 kB",
  "originSize": "2.07 MB",
}
`;

exports[`woff2 > woff2 1`] = `
{
  "hash": "d674312638c499b7137a116536c1019cb8a263e16fe1f2561cf68316535cfde3",
  "newSize": "3.50 kB",
  "originSize": "942.54 kB",
}
`;

exports[`woff2 > woff2 2 eot 1`] = `
{
  "hash": "2486ff4cbc1b1bad41b2dfa7088fe69125c52a7e922337fe93caf2130ae7215c",
  "newSize": "6.83 kB",
  "originSize": "942.54 kB",
}
`;

exports[`woff2 > woff2 2 ttf 1`] = `
{
  "hash": "710937d80c717c24fd5636deae82feb1f71d0c134a2e7d15a968785259f7711c",
  "newSize": "6.63 kB",
  "originSize": "942.54 kB",
}
`;

exports[`woff2 > woff2 2 woff 1`] = `
{
  "hash": "d30834d9f45cbfb6c7fc9f2e578636d47efafb3dbfedbc188c01bfdcb393f96d",
  "newSize": "6.70 kB",
  "originSize": "942.54 kB",
}
`;
